pkg/jit/jitdump.go

```diff
@@ -0,0 +1,460 @@
+	Lineno  uint32 // source file line number (starting at 1)
+	Discrim uint32 // column discriminator, 0 is default
+	Name    string // source file name in ASCII
+}
+
+const debugEntryFixedSize int = 16 // size of DebugEntry fixed-sized fields
+
+// JRCodeDebugInfo represents a JITCodeDebugInfo record.
+type JRCodeDebugInfo struct {
+	Prefix   *JRPrefix     // the record header
+	CodeAddr uint64        // address of function for which the debug information is generated
+	NREntry  uint64        // number of debug entries for the function
+	Entries  []*DebugEntry // array of NREntry debug entries for the function
+}
+
+const jrCodeDebugInfoFixedSize int = jrPrefixSize + 16 // size of JRCodeDebugInfo fixed-sized fields
+
+// JRCodeUnwindingInfo represents a JITCodeUnwindingInfo record.
+type JRCodeUnwindingInfo struct {
+	Prefix         *JRPrefix // the record header
+	UnwindingSize  uint64    // the size in bytes of the unwinding data table at the end of the record
+	EHFrameHDRSize uint64    // the size in bytes of the DWARF EH Frame Header at the start of the unwinding data table at the end of the record
+	MappedSize     uint64    // the size of the unwinding data mapped in memory
+	UnwindingData  []byte    // an array of unwinding data, consisting of the EH Frame Header, followed by the actual EH Frame
+}
+
+const jrCodeUnwindingInfoFixedSize int = jrPrefixSize + 24 // size of JRCodeUnwindingInfo fixed-sized fields
+
+// JITDump represents the loaded jitdump.
+type JITDump struct {
+	Header        *JITHeader             // the jitdump file header
+	CodeLoads     []*JRCodeLoad          // JITCodeLoad records
+	CodeMoves     []*JRCodeMove          // JITCodeMove records
+	DebugInfo     []*JRCodeDebugInfo     // JITCodeDebugInfo records
+	UnwindingInfo []*JRCodeUnwindingInfo // JITCodeUnwindingInfo records
+}
+
+// jitDumpParser is used to parse a jitdump file.
+type jitDumpParser struct {
+	logger     log.Logger
+	buf        *bufio.Reader
+	endianness binary.ByteOrder
+}
+
+// newParser initializes a jitDumpParser.
+func newParser(logger log.Logger, rd io.Reader) (*jitDumpParser, error) {
+	p := &jitDumpParser{
+		logger: logger,
+	}
+
+	p.buf = bufio.NewReader(rd)
+
+	magic, err := p.buf.Peek(4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read magic number from jitdump file: %w", err)
+	}
+
+	switch {
+	case bytes.Equal(magic, []byte{'J', 'i', 'T', 'D'}):
+		p.endianness = binary.BigEndian
+	case bytes.Equal(magic, []byte{'D', 'T', 'i', 'J'}):
+		p.endianness = binary.LittleEndian
+	default:
+		return nil, fmt.Errorf("%w: %#x", err, magic)
+	}
+
+	return p, nil
+}
+
+// isUnexpectedIOError ensures all EOF are unexpected EOF errors.
+func isUnexpectedIOError(err error) error {
+	if errors.Is(err, io.EOF) {
+		return io.ErrUnexpectedEOF
+	}
+	return err
+}
+
+// read reads structured binary data from the jitdump file into data.
+func (p *jitDumpParser) read(data any) error {
+	return binary.Read(p.buf, p.endianness, data)
```

Comment:
`binary.Read` (and Write) [allocate even on the fast path](https://github.com/golang/go/blob/3c7884a21b2929b5a42d6a1f237f5fee2ae72ba8/src/encoding/binary/binary.go#L232). While this might not seem like a problem, it will create a lot of short-lived allocations increase pressure on the GC even more.

We spend way too many cycles in the Garbage Collector already (~30%) and want to reduce this. I recently added even more `binary.Write`s and I am planning to remove them soon with more efficient low level bits, `binary.LittleEndian.Put*`.

Perhaps here we could use `binary.LittleEndian. Uint*`?

Reply:
Do you mean [`binary.ReadUvarint`](https://pkg.go.dev/encoding/binary#ReadUvarint)? Yes, that could be an optimization, but correct if I am wrong, it would be for `uint64` fields only

Reply:
More like `p.endianness.Uint32|64(p.buf)` as what's currently the caller of this function knows exactly what type it wants to read. But the way `binary.Read` is written it will always perform allocations even when not necessary (and it's not necessary for reading integers).

Reply:
Ah I see the [`ByteOrder` interface](https://pkg.go.dev/encoding/binary#ByteOrder), it looks like it would still require to allocate a slice of bytes, they do not take an `io.Reader`

it is what `binary.Read` does: https://cs.opensource.google/go/go/+/refs/tags/go1.20:src/encoding/binary/binary.go;l=254

Reply:
Sorry for clarification, it does, because the size cannot be predicted, and therefore is allocated on the heap. Whereas for uint64, a `make([]byte, 4)` is not allocated on the heap as the compiler knows exactly the size and can manage it on the stack.

Reply:
> Line 254 doesn't cause the allocation, line 232 does: [cs.opensource.google/go/go/+/refs/tags/go1.20:src/encoding/binary/binary.go;l=232](https://cs.opensource.google/go/go/+/refs/tags/go1.20:src/encoding/binary/binary.go;l=232)

Yes

---

Stack vs Heap: I see now, thank you for the explanation! I can easily observe the difference now:

```console
$ cat >main.go <<EOF
package main

func main() {
        n := 8
        _ = make([]byte, n)
        _ = make([]byte, 8)
}
EOF

$ go build -gcflags=-m main.go
# command-line-arguments
./main.go:3:6: can inline main
./main.go:5:10: make([]byte, n) escapes to heap
./main.go:6:10: make([]byte, 8) does not escape
```